qpid/cpp/src/qpid/broker/amqp/Outgoing.cpp

```diff
@@ -79,6 +79,9 @@ OutgoingFromQueue::OutgoingFromQueue(Broker& broker, const std::string& source,
 
 void OutgoingFromQueue::init()
 {
+    // observe queue for changes to track undeliverable messages
+    queue->getObservers().add(boost::shared_ptr<QueueObserver>(shared_from_this()));
+
```

Comment:
Either of those works. One other point is that we need to deregister the observer when the consumer ends.